# Iris Species Classification

## Project Overview

The **Iris dataset** is one of the most famous datasets in the field of machine learning. It was originally used in R.A. Fisher's classic 1936 paper *The Use of Multiple Measurements in Taxonomic Problems*. This dataset contains 150 samples from three different species of Iris flowers: *Iris-setosa*, *Iris-versicolor*, and *Iris-virginica*, with four features recorded for each sample.

The objective of this project is to build a classification model that predicts the species of an Iris flower based on the given features, using machine learning techniques.

## Objective

The goal of this project is to:

1. Understand and visualize the dataset.
2. Train a classification model to accurately predict the species of Iris flowers based on the four features.
3. Evaluate the performance of different classification models using metrics such as accuracy, precision, recall, and F1-score.

## Dataset Description

The dataset includes the following columns:

- **Id**: Unique identifier for each sample.
- **SepalLengthCm**: Sepal length of the flower in centimeters.
- **SepalWidthCm**: Sepal width of the flower in centimeters.
- **PetalLengthCm**: Petal length of the flower in centimeters.
- **PetalWidthCm**: Petal width of the flower in centimeters.
- **Species**: The species of the flower (*Iris-setosa*, *Iris-versicolor*, or *Iris-virginica*).

### Summary:

- **Number of Observations**: 150
- **Number of Features**: 5 (4 features + 1 target variable)

### Problem Statement

- **Model Training**: Train the machine learning model with the data so that it can able to predict the class of Iris species.
- **Model Evaluation**: The objective of model evaluation is to evaluate the performance of the trained ML model using different evaluation metrics such as accuracy, presicion, recal and F1 scores.
- **Model Optimization** Optimize the performance of the trained ML model using cross validation and hyperparameter tuning so that it can predict the class of Iris species more accurately.



### Load Libraries

In [1]:
# General
import pandas as pd
import numpy as np
import os
import warnings
import pickle

# Preprocessing
from sklearn.model_selection import train_test_split

# Model and Evaluation Metrics
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

### Settings

In [2]:
# Warning
warnings.filterwarnings("ignore")
# Path
data_path = "../data"
model_path = "../models"
csv_path = os.path.join(data_path, "Iris_cleaned.csv")

### Load Data

In [3]:
df = pd.read_csv(csv_path)

In [4]:
# Check data
df.head()

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0            5.1           3.5            1.4           0.2        0
1            4.9           3.0            1.4           0.2        0
2            4.7           3.2            1.3           0.2        0
3            4.6           3.1            1.5           0.2        0
4            5.0           3.6            1.4           0.2        0

### Preprocessing

In [6]:
# Separate the Input and Output Features
X = df.iloc[:, :-1]
y = df["Species"]

In [25]:
# Split training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.4, random_state= 42)

### Model Training and Evaluation

In [26]:
# Define a function to train the model and evaluate
def train_evaluate(model):
    # Train the model
    model.fit(X_train, y_train)

    # Make prediction on training and testing data
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Print model evaluation scores
    print("=" * 60)
    print("EVALUATION METRICS FOR TRAINING")
    print("=" * 60)
    print(f"Accuracy: {accuracy_score(y_train, y_train_pred): 0.2f}")
    print(f"Precision: {precision_score(y_train, y_train_pred, average='weighted'): 0.2f}")
    print(f"Recall: {recall_score(y_train, y_train_pred, average='weighted'): 0.2f}")
    print(f"F1: {f1_score(y_train, y_train_pred, average='weighted'): 0.2f}")
    print("=" * 60)
    print("EVALUATION METRICS FOR TESTING")
    print("=" * 60)
    print(f"Accuracy: {accuracy_score(y_test, y_test_pred): 0.2f}")
    print(f"Precision: {precision_score(y_test, y_test_pred, average='weighted'): 0.2f}")
    print(f"Recall: {recall_score(y_test, y_test_pred, average='weighted'): 0.2f}")
    print(f"F1: {f1_score(y_test, y_test_pred, average='weighted'): 0.2f}")

In [27]:
# Try with XGBoost Classifier
xgbc = XGBClassifier()
train_evaluate(xgbc)

EVALUATION METRICS FOR TRAINING
Accuracy:  1.00
Precision:  1.00
Recall:  1.00
F1:  1.00
EVALUATION METRICS FOR TESTING
Accuracy:  0.98
Precision:  0.98
Recall:  0.98
F1:  0.98


### Conclusion

The model's performance on the training data is very high, with an accuracy of **100%**. This suggests that the model is doing an excellent job of predicting the outcomes for the training dataset. The precision of **100%** indicates that all of the predicted species are classified collectly, and the recall of **100%** suggests that the model is able to correctly identify all species. The F1-score of **100%**, which is the harmonic mean of precision and recall, also shows a strong balance between them.

On the testing data, the model’s accuracy is also excellent which is **98%**, which is very close the training accuracy. The precision (**98%**) and recall (**98%**) also suggest that most of the species are correctly classified(a vary few false positive and false negetive are present). This indicates that the model is very much effective when dealing with unseen data and generalize the patterns it learned during training.

- **Precision**: The value of **0.98** means that when the model predicts species correctly, **98%** of the time, they actually are correct species. The model makes a very false positive errors compared to the training data.
- **Recall**: The value of **0.98** means that the model is able to identify **98%** of the actual species. It’s missing a very few species (false negatives), meaning it's very good in its predictions.
- **F1-Score**: This is a balance between precision and recall. While the model is performing very well.